### PR TITLE
Fix bug in HighLevelGraph.cull

### PR DESCRIPTION
**Pull request by rjzamora**
As discussed in https://github.com/dask/dask/issues/8991, calling `HighLevelGraph.cull` seems to produce a `key_dependencies` attribute that is inconsistent with manually-computed key dependencies (dependencies computed **without** an optimization pass).  This is because `HighLevelGraph.cull` incorrectly conflates Layer dependencies with key dependencies in the result of `Layer.cull`.  This PR attempts to rectify this error in a simple way.

### CODE REVIEW

File: dask/highlevelgraph.py [hunk 1]
```diff
@@ -931,6 +931,8 @@ def cull(self, keys: Iterable) -> HighLevelGraph:
         hlg: HighLevelGraph
             Culled high level graph
         """
+        from dask.layers import Blockwise
```
Comment (rjzamora) [inline dask/highlevelgraph.py:934]: I'd rather not import Blockwise like this, but it also feels wasteful not to use culled_deps for the most common case where the layer and key dependencies will be the same. Open to suggestions here.

File: dask/tests/test_distributed.py [hunk 2]
```diff
@@ -794,7 +794,7 @@ def test_set_index_no_resursion_error(c):
         pytest.fail("dd.set_index triggered a recursion error")
 
 
-@pytest.mark.xfail(reason="https://github.com/dask/dask/issues/8991", strict=True)
+# @pytest.mark.xfail(reason="https://github.com/dask/dask/issues/8991", strict=True)
```
Comment (rjzamora) [inline dask/tests/test_distributed.py:797]: @ian-r-rose - Do you think we should remove the xfail here, or just remove the test altogether?

Comment (ian-r-rose) [reply dask/tests/test_distributed.py:797]: I'm torn on this -- I'd love to have some test that probes whether `key_dependencies` are correct, but I really don't like this one. Do you have any ideas for how to improve it?

Comment (ian-r-rose) [reply dask/tests/test_distributed.py:797]: What I don't like about this is that it is pretty far from user code, and calls a bunch of private methods.

> Maybe we really don't need this to be in test_distributed.py? That is, in test_layers.py we can just test that calling cull on various layers does not change the output of graph.get_all_dependencies()?

I like this idea better. In my memory, the last time I tried this I ran into problems with the fact that actual dependency logic happens in the pack/unpack process. But I could have that wrong, I haven't looked in detail recently.

Comment (rjzamora) [reply dask/tests/test_distributed.py]: >I ran into problems with the fact that actual dependency logic happens in the pack/unpack process. But I could have that wrong, I haven't looked in detail recently.

Yes - The **actual** dependency logic does happen in the unpack process. However, when `key_dependencies` is empty, `get_all_dependencies` will manually crawl through the materialized graph and use `keys_in_tasks` to construct the "ground truth" for `key_dependencies`. Therefore, it think it should be a good way to test that `HighLevelGraph.cull` isn't tainting the result.

**Comment by ian-r-rose** [comment]
Thanks for this @rjzamora! Your test is much better than my old one.

I think it should also be safe to revert the workaround in `compute_as_if_collection` in #8998 as well (I can try to verify)

**Comment by ian-r-rose** [comment]
Update: I've manually confirmed that it is safe to remove the workaround from #8998 with this fix. I ran the following locally on my laptop, which fails with poor memory usage when the deps are bad:

```python
import dask.datasets
import distributed
import dask.dataframe as dd

client= distributed.Client()

ddf = dask.datasets.timeseries(freq="100ms", partition_freq="1D")
ddf.to_parquet("tseries", compute=True)
ddf = dd.read_parquet("tseries").shuffle("x")

ddf.to_parquet("tseries2", compute=True)
```

**Review by ian-r-rose — APPROVED** [review]
Thanks @rjzamora!

**Comment by rjzamora** [comment]
Thanks for the review @ian-r-rose !

**Pull request merged**